server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/JmapChangeFactory.java

```diff
@@ -0,0 +1,292 @@
+                .build();
+
+            Stream<JmapChange> shareeMailboxChanges = getSharees(messageAdded.getMailboxId(), messageAdded.getUsername(), mailboxManager)
+                .flatMap(name -> Stream.of(
+                    MailboxChange.builder()
+                        .accountId(AccountId.fromString(name))
+                        .state(stateFactory.generate())
+                        .date(now)
+                        .isCountChange(true)
+                        .updated(ImmutableList.of(messageAdded.getMailboxId()))
+                        .delegated()
+                        .build(),
+                    EmailChange.builder()
+                        .accountId(AccountId.fromString(name))
+                        .state(stateFactory.generate())
+                        .date(now)
+                        .isDelegated(true)
+                        .created(messageAdded.getMessageIds())
+                        .build()));
+
+            return Stream.concat(Stream.of(ownerMailboxChange, ownerEmailChange), shareeMailboxChanges)
+                .collect(Guavate.toImmutableList());
+        }
+        if (event instanceof MailboxListener.FlagsUpdated) {
+            MailboxListener.FlagsUpdated messageFlagUpdated = (MailboxListener.FlagsUpdated) event;
+            boolean isSeenChanged = messageFlagUpdated.getUpdatedFlags()
+                .stream()
+                .anyMatch(flags -> flags.isChanged(Flags.Flag.SEEN));
+
+            EmailChange ownerEmailChange = EmailChange.builder()
+                .accountId(AccountId.fromUsername(messageFlagUpdated.getUsername()))
+                .state(stateFactory.generate())
+                .date(now)
+                .isDelegated(false)
+                .updated(messageFlagUpdated.getMessageIds())
+                .build();
+
+            Stream<EmailChange> shareeEmailChanges =  getSharees(messageFlagUpdated.getMailboxId(), messageFlagUpdated.getUsername(), mailboxManager)
+                .map(name -> EmailChange.builder()
+                    .accountId(AccountId.fromString(name))
+                    .state(stateFactory.generate())
+                    .date(now)
+                    .isDelegated(true)
+                    .updated(messageFlagUpdated.getMessageIds())
+                    .build());
+
+            if (isSeenChanged) {
+                MailboxChange ownerMailboxChange = MailboxChange.builder()
+                    .accountId(AccountId.fromUsername(messageFlagUpdated.getUsername()))
+                    .state(stateFactory.generate())
+                    .date(now)
+                    .isCountChange(true)
+                    .updated(ImmutableList.of(messageFlagUpdated.getMailboxId()))
+                    .build();
+
+                Stream<JmapChange> shareeChanges = Stream.concat(shareeEmailChanges,
+                    getSharees(messageFlagUpdated.getMailboxId(), messageFlagUpdated.getUsername(), mailboxManager)
```

Comment:
Unfortunately because of the if condition it's becoming too complex to try to get the sharees only once for FlagsUpdated event. Maybe there is a nice simple solution, don't hesitate  to give feedback if you have an idea thanks